pkg/promtail/targets/objectstore/object_reader.go

```diff
@@ -0,0 +1,245 @@
+package objectstore
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/promtail/positions"
+)
+
+type objectReader struct {
+	storeName    string
+	handler      api.EntryHandler
+	objectClient chunk.ObjectClient
+	logger       log.Logger
+	readerMtx    sync.RWMutex
+	labels       model.LabelSet
+	positions    positions.Positions
+	quit         chan struct{}
+	done         chan struct{}
+	object       *promtailObject
+	active       bool
+}
+
+type promtailObject struct {
+	Key        string
+	ModifiedAt time.Time
+	Size       int64
+	BytesRead  int64
+}
+
+func newObjectReader(storeName string, object chunk.StorageObject, objectClient chunk.ObjectClient, logger log.Logger, labels model.LabelSet, positions positions.Positions, handler api.EntryHandler, pos int64) (*objectReader, error) {
+	objectReader := &objectReader{
+		storeName:    storeName,
+		logger:       logger,
+		labels:       labels,
+		positions:    positions,
+		handler:      handler,
+		quit:         make(chan struct{}),
+		done:         make(chan struct{}),
+		objectClient: objectClient,
+		active:       true,
+		object: &promtailObject{
+			Key:        object.Key,
+			ModifiedAt: object.ModifiedAt,
+			BytesRead:  pos,
+			Size:       0,
+		},
+	}
+
+	go objectReader.run()
+	ObjectsActive.Add(1.)
+	return objectReader, nil
+}
+
+func (r *objectReader) run() {
+	level.Info(r.logger).Log("msg", "start reading object", "object", r.object.Key)
+	bufReader, err := r.GetObject()
+	if err != nil {
+		level.Error(r.logger).Log("msg", "error in fetching and initializing object reader", "err", err)
+		r.readerMtx.Lock()
+		r.active = false
+		r.readerMtx.Unlock()
```

Comment:
You could use an atomic boolean instead. Checkout the [Cloudflare target](https://github.com/grafana/loki/blob/main/clients/pkg/promtail/targets/cloudflare/target.go#L49).

Reply:
Sure. Thanks, using atomic boolean in updated code